blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ElasticTokensIO.java

```
@@ -0,0 +1,291 @@
+
+    public static final String ES_DOCUMENT_TYPE = "tokens";
+
+    protected final Histogram batchHistogram = Metrics.histogram(getClass(), "Batch Sizes");
+    protected final Timer writeTimer = Metrics.timer(getClass(), "Write Duration");
+
+    protected final Timer esMetricNamesQueryTimer = Metrics.timer(getClass(), "ES Metric Names Query Duration");
+
+    private static final Logger log = LoggerFactory.getLogger(ElasticTokensIO.class);
+
+    public static String ELASTICSEARCH_TOKEN_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_TOKEN_INDEX_NAME_WRITE);
+    public static String ELASTICSEARCH_TOKEN_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_TOKEN_INDEX_NAME_READ);
+
+    private Client client;
+
+    public ElasticTokensIO() {
+        this(RemoteElasticSearchServer.getInstance());
+    }
+
+    public ElasticTokensIO(Client client) {
+        this.client = client;
+    }
+
+    public ElasticTokensIO(ElasticClientManager manager) {
+        this(manager.getClient());
+    }
+
+    @Override
+    public void insertDiscovery(Token token) throws IOException {
+        List<Token> batch = new ArrayList<>();
+        batch.add(token);
+        insertDiscovery(batch);
+    }
+
+    @Override
+    public void insertDiscovery(List<Token> tokens) throws IOException {
+        batchHistogram.update(tokens.size());
+        if (tokens.size() == 0) return;
+
+        Timer.Context ctx = writeTimer.time();
+        try {
+
+            BulkRequestBuilder bulk = client.prepareBulk();
+
+            for (Token token : tokens) {
+                bulk.add(createSingleRequest(token));
+            }
+
+            bulk.execute().actionGet();
+        } catch (EsRejectedExecutionException esEx) {
+            log.error(("Error during bulk insert to ES with status: [" + esEx.status() + "] " +
+                    "with message: [" + esEx.getDetailedMessage() + "]"));
```

We don't want to bubble up the exception?